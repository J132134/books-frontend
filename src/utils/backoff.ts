export class CancelledError extends Error {
  constructor(public inner?: Error) {
    super('Retry cancelled');

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, CancelledError);
    }
    this.name = this.constructor.name;
  }
}

interface ExpBackoffOptions {
  maxRetries?: number;
  backoffTimeUnit?: number;
}

const DEFAULT_MAX_RETRIES = Infinity;
const DEFAULT_TIME_UNIT = 100;

export function runWithExponentialBackoff<T>(
  factory: () => Promise<T>,
  options?: ExpBackoffOptions,
): Promise<T> {
  const {
    maxRetries = DEFAULT_MAX_RETRIES,
    backoffTimeUnit = DEFAULT_TIME_UNIT,
  } = options || {};
  let count = 0;
  function runOnce(): Promise<T> {
    return factory().catch((err) => {
      if (err instanceof CancelledError) {
        return Promise.reject(err);
      }
      const window = 2 ** count;
      count += 1;
      if (count >= maxRetries) {
        return Promise.reject(err);
      }
      const windowIdx = Math.floor(Math.random() * window) + window;
      return new Promise((resolve) => setTimeout(resolve, windowIdx * backoffTimeUnit))
        .then(runOnce);
    });
  }
  return runOnce();
}
