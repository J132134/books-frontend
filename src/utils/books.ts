import * as BookApi from 'src/types/book';

/* 미완결 된 Series 도서중 공개(opened) 된 마지막 회차의 표지 ID를 구함 */
export function getThumbnailIdFromBookDetail(book: BookApi.Book | null) {
  if (!book) {
    return null;
  }
  if (book.series) {
    if (
      !book.series.property.is_completed && book.series.property.opened_last_volume_id.length !== 0
    ) {
      return book.series.property.opened_last_volume_id;
    }
  }
  return book.id;
}
