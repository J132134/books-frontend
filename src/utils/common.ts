import { notifySentry } from './sentry';

// tslint:disable-next-line
export const safeJSONParse = (source: string | null, defaultValue: any) => {
  if (!source) {
    return defaultValue;
  }
  try {
    return JSON.parse(source);
  } catch (err) {
    notifySentry(err);
    return defaultValue;
  }
};

export const isJSON = (source: string) => {
  try {
    JSON.parse(source);
    return true;
  } catch (err) {
    return false;
  }
};

export const timeAgo = (timestamp: number) => {
  let diff = Date.now() - timestamp;

  // 분단위로 변환
  diff = Math.floor(diff / (60 * 1000));

  if (diff < 1) {
    return '방금 전';
  }

  if (diff < 60) {
    return `${diff}분 전`;
  }

  // 시간단위로 변환
  diff = Math.floor(diff / 60);
  if (diff < 24) {
    return `${diff}시간 전`;
  }

  // 일 단위로 변환
  diff = Math.floor(diff / 24);
  if (diff < 30) {
    return `${diff}일 전`;
  }

  // 월 단위로 변환
  diff = Math.floor(diff / 30);

  return `${diff}개월 전`;
};
