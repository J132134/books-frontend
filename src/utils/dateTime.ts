export function createTimeLabel() {
  const date = new Date();
  date.setHours(
    date.getHours()
    + date.getTimezoneOffset() / 60
    + 9,
  );
  const hour = date
    .getHours()
    .toString()
    .padStart(2, '0');
  const min = date
    .getMinutes()
    .toString()
    .padStart(2, '0');

  return `${hour}시 ${min}분`;
}
