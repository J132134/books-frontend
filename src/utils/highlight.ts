import React from 'react';

// eslint-disable-next-line
export const getEscapedString = (source: string) => {
  return source
    .replace(/</giu, '&lt;')
    .replace(/>/giu, '&gt;')
    .replace(/&lt;strong([^&]*)&gt;(.*?)&lt;\/strong&gt;/giu, '<strong$1>$2</strong>');
};

/**
 * 줄바꿈을 React <br />로 치환합니다.
 */
export function newlineToReactNode(value: string): React.ReactNode {
  const lines = value.split(/\r\n|\r|\n/g);
  const nodes = lines
    .map((line, idx) => (
      [idx !== 0 && React.createElement('br', { key: String(idx) }), line]
    ));
  return [].concat(...nodes);
}
