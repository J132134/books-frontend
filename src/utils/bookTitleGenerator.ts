import * as BookApi from 'src/types/book';
import { getEscapedString } from 'src/utils/highlight';

export const bookTitleGenerator = (book: BookApi.Book) => {
  if (!book) {
    return '';
  }
  if (book.series) {
    if (book.title.prefix) {
      return getEscapedString(`${book.title.prefix} ${book.series.property.title}`);
    }
    return getEscapedString(book.series.property.title);
  }
  if (book.title) {
    if (book.title.prefix) {
      return getEscapedString(`${book.title.prefix} ${book.title.main}`);
    }
  }
  return getEscapedString(book.title.main);
};
