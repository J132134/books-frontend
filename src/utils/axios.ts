import axios, { AxiosInstance } from 'axios';
import { AbortError } from 'p-retry';
import { tokenInterceptor } from 'src/utils/axiosInterceptors';

export enum OAuthRequestType {
  NORMAL = 'NORMAL', // 실패하더라도 특별한 행동을 하지 않는다
  CHECK = 'CHECK', // rt 갱신 시도, at 확인 후 실패하더라도 실패한 결과를 반환
  STRICT = 'STRICT', // rt 갱신, at 확인 후 로그인되지 않았다면 로그인 페이지로 이동
}

export interface CustomRequestConfig {
  authorizationRequestType: OAuthRequestType;
}

export const { CancelToken } = axios;

// eslint-disable-next-line no-process-env
const TIME_OUT = 7000;

const globalAxiosConfig = () => {
  axios.defaults.timeout = TIME_OUT;
};
globalAxiosConfig();

export function wrapCatchCancel<F extends Function>(f: F): F {
  return (function (...args: any[]) {
    try {
      // @ts-ignore
      return f.apply(this, args);
    } catch (err) {
      if (axios.isCancel(err)) {
        // 취소되었다면 재시도하지 않음
        throw new AbortError(err);
      }
      throw err;
    }
  } as unknown) as F;
}

const createAxiosInstances = (): AxiosInstance => {
  const instance = axios.create({ timeout: TIME_OUT });
  instance.interceptors.response.use((onFulfilled) => onFulfilled, tokenInterceptor);
  if (process.env.IS_SERVER) {
    // 여기서는 Node.js가 로드한다는 걸 알고 있으므로 require를 직접 씁니다.
    // eslint-disable-next-line global-require
    const AxiosLogger = require('axios-logger');
    instance.interceptors.request.use(AxiosLogger.requestLogger, AxiosLogger.errorLogger);
    instance.interceptors.response.use(
      (res) => AxiosLogger.responseLogger(res, { data: false }),
      AxiosLogger.errorLogger,
    );
  }
  return instance;
};
export default createAxiosInstances();
