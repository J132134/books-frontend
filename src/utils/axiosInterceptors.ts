import axios, { AxiosError } from 'axios';

export const authorizeTokenFallback = async (rejectedError: AxiosError) => {
  await axios.get('/ridi/authorize/', {
    withCredentials: true,
    baseURL: process.env.NEXT_STATIC_ACCOUNT_API,
    params: {
      client_id: process.env.NEXT_STATIC_RIDI_OAUTH2_CLIENT_ID,
      response_type: 'code',
      redirect_uri: `${process.env.NEXT_STATIC_ACCOUNT_API}/ridi/complete`,
    },
  });

  // 원본 요청
  return axios(rejectedError.config);
};

export const refreshTokenFallback = async (rejectedError: AxiosError) => {
  try {
    await axios.post('/ridi/token', null, {
      baseURL: process.env.NEXT_STATIC_ACCOUNT_API,
      withCredentials: true,
    });
  } catch (error) {
    const { response } = error;
    if (!response) {
      return Promise.reject(error);
    }
    // Access-Token 확인
    return authorizeTokenFallback(rejectedError);
  }

  // 원본 요청
  return axios(rejectedError.config);
};

export const tokenInterceptor = (onRejected: AxiosError) => {
  const { response } = onRejected;
  if (!response) {
    return Promise.reject(onRejected);
  }

  const tokenUrl = `${process.env.NEXT_STATIC_ACCOUNT_API}/ridi/token/`;
  if (
    (response.status === 401 || response.status === 403)
    && response.config.url !== tokenUrl
  ) {
    return refreshTokenFallback(onRejected);
  }

  return Promise.reject(onRejected);
};
