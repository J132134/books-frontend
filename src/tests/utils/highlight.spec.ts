import { render, cleanup } from '@testing-library/react';

import { getEscapedNode } from 'src/utils/highlight';

afterEach(() => cleanup());

describe('getEscapedNode', () => {
  it.each([
    ['normal text with <angular brackets>', 'normal text with <angular brackets>'],
    ['some <strong>bold</strong> texts', 'some bold texts'],
    ['nes<strong>ted<strong> bold</strong> tex</strong>ts', 'nested bold texts'],
    ['tag <strong class="foo">with class</strong>', 'tag with class'],
  ])('should escape correctly', (source, plainText) => {
    const renderResult = render(getEscapedNode(source));
    expect(renderResult.container).toMatchSnapshot();
    expect(renderResult.container.textContent).toEqual(plainText);
  });
});
