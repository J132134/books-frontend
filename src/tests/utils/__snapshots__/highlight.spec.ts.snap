// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEscapedNode should escape correctly 1`] = `
<div>
  normal text with &lt;angular brackets&gt;
</div>
`;

exports[`getEscapedNode should escape correctly 2`] = `
<div>
  some 
  <strong>
    bold
  </strong>
   texts
</div>
`;

exports[`getEscapedNode should escape correctly 3`] = `
<div>
  nes
  <strong>
    ted
    <strong>
       bold
    </strong>
     tex
  </strong>
  ts
</div>
`;

exports[`getEscapedNode should escape correctly 4`] = `
<div>
  tag 
  <strong>
    with class
  </strong>
  
</div>
`;
